src/coreclr/jit/morph.cpp

```diff
@@ -6452,24 +6483,39 @@ void Compiler::fgValidateIRForTailCall(GenTreeCall* call)
         }
     };
 
+    JITDUMP("Validating IR for tail call candidate [%06u] in " FMT_STMT "\n", dspTreeID(call), compCurStmt->GetID());
+
     TailCallIRValidatorVisitor visitor(this, call);
-    for (Statement* stmt = compCurStmt; stmt != nullptr; stmt = stmt->GetNextStmt())
+    for (Statement* stmt = compCurStmt; visitor.IsValid() && (stmt != nullptr); stmt = stmt->GetNextStmt())
     {
         visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
```

Comment:
Nit: I think it would be a bit cleaner to return `WALK_ABORT` on validation failure in the visitor (looks like at least `GT_RETURN` needs to be fixed) and have early exits for when `WalkTree` returns that.